programs/axelar-solana-gateway/src/state/signature_verification.rs

```diff
@@ -353,4 +359,97 @@ mod tests {
         // Hash should remain unchanged after failure
         assert_eq!(verification.signing_verifier_set_hash, initial_hash);
     }
+
+    #[test]
+    fn test_check_slot_is_done_returns_slot_already_verified_error() {
+        let mut verification = SignatureVerification::default();
+
+        // Set the bit at position 0 to indicate it's already verified
+        verification.signature_slots[0] = 0b0000_0001;
+
+        let verifier_leaf = VerifierSetLeaf {
+            signer_pubkey: PublicKey::Secp256k1([0; 33]),
+            signer_weight: 1,
+            position: 0u8.into(),
+            quorum: 1,
+            set_size: 1u8.into(),
+            domain_separator: [0; 32],
+            nonce: 0,
+        };
+
+        assert_eq!(
+            verification.check_slot_is_done(&verifier_leaf),
+            Err(GatewayError::SlotAlreadyVerified)
+        );
```

Comment:
I would add another assertion here, assuming the verification is idempotent/stateless

```suggestion
        let verifier_leaf = VerifierSetLeaf {
            signer_pubkey: PublicKey::Secp256k1([0; 33]),
            signer_weight: 1,
            position: 0u8.into(),
            quorum: 1,
            set_size: 1u8.into(),
            domain_separator: [0; 32],
            nonce: 0,
        };

        assert_eq!(
            verification.check_slot_is_done(&verifier_leaf),
            Ok(())
        );
        
        // Set the bit at position 0 to indicate it's already verified
        verification.signature_slots[0] = 0b0000_0001;

        assert_eq!(
            verification.check_slot_is_done(&verifier_leaf),
            Err(GatewayError::SlotAlreadyVerified)
        );
```

Reply:
Looking at the other test below I guess the slot mask change wouldn't be needed then?

Reply:
If I understand this correctly, I think we're not calling `process_signature` in this test (which flips the bit as a side effect), but flipping it manually here.

Reply:
Yes the these two tests are complementary and my suggestion would make this one redundant. I'll close this thread.